server/token.go

```diff
@@ -179,29 +179,28 @@ func (s *IDPServer) handleAuthorizationCodeGrant(w http.ResponseWriter, r *http.
 	}
 	s.mu.Unlock()
 	if !ok {
-		writeTokenEndpointError(w, http.StatusBadRequest, "invalid_grant", "code not found")
+		writeHTTPError(w, r, http.StatusBadRequest, ecInvalidGrant, "code not found", nil)
 		return
 	}
 	if httpStatusCode, err := ar.allowRelyingParty(r); err != nil {
-		//log.Printf("XXX Error allowing relying party: %v", err)
-		writeTokenEndpointError(w, httpStatusCode, "invalid_client", err.Error())
+		writeHTTPError(w, r, httpStatusCode, ecInvalidClient, "client authentication failed", err)
 		return
 	}
 	if ar.RedirectURI != r.FormValue("redirect_uri") {
-		writeTokenEndpointError(w, http.StatusBadRequest, "invalid_grant", "redirect_uri mismatch")
+		writeHTTPError(w, r, http.StatusBadRequest, ecInvalidGrant, "redirect_uri mismatch", nil)
```

Comment:
optional: consider logging the URI for debugging
```suggestion
		writeHTTPError(w, r, http.StatusBadRequest, ecInvalidGrant, "redirect_uri mismatch", fmt.Errorf("got redirect_uri %q, want %q", r.FormValue("redirect_uri"), ar.RedirectURI))
```